### PR TITLE
Reduce stop impact cost when flow data is present

### PR DESCRIPTION
**Pull request by gknisely**
# Issue

This PR(collaboration between @dgearhart  and @gknisely ) is the first attempt at reducing the favoring of longer routes or highways for very little time improvement.  Basically, we reduced transition cost when edge has flow information.  A bit has been added to the edgelabel which indicates that we have flow information or traffic set from one of the flow sources: kFreeFlowMask, kConstrainedFlowMask, kPredictedFlowMask, or kCurrentFlowMask.

Note: This update is for only the auto costing for now.   

## Examples
| BEFORE | AFTER |
|---|---|
| image | image |
| image | image |
| image | image |

## Performance
No performance impact with these changes
### Average cross country route
| MASTER | BRANCH |
|---|---|
| 219 ms | 219 ms |

## Tasklist

 - [x] Add tests
 - [x] Add #fixes with the issue number that this PR addresses
 - [x] Update the [changelog](CHANGELOG.md)

### CODE REVIEW

**Comment by dnesbitt61** [comment]
can you provide a description of the reason for the changes and the expected impact?

**Comment by gknisely** [comment]
> can you provide a description of the reason for the changes and the expected impact?

Yes.  Still in draft mode for this PR

File: src/sif/autocost.cc
Comment (gknisely) [inline src/sif/autocost.cc:276]: This is a TODO:  We should try to update TransitionCostReverse with `const EdgeLabel& pred` vs `const baldr::DirectedEdge* pred`; however, we did not have time to make this change.  

@dnesbitt noted that when expanding in reverse the edge added to the edge label is in the direction of expansion.  This is the reverse of the direction of travel. This is why the opposing predecessor directed edge is passed into the TransitionCostReverse method rather than the predecessor EdgeLabel. With some thought and care it may be possible to refactor this but I'm not sure. We need to make sure idx is computed correctly is one complexity

File: src/sif/autocost.cc
Comment (gknisely) [inline src/sif/autocost.cc:112]: Ported this over from the motorcycle costing.

File: src/sif/autocost.cc
Comment (gknisely) [inline src/sif/autocost.cc:366]: Again, ported from motorcycle costing

File: src/sif/autocost.cc
Comment (gknisely) [inline src/sif/autocost.cc:454]: Adding `&flow_sources` here allows us to save off which speed sources were used in this speed calculation.  This is flattened down to a single bit on the edgelabel.  Which basically indicates if we have some sort of traffic data for this edge.

File: src/thor/bidirectional_astar.cc
Comment (gknisely) [inline src/thor/bidirectional_astar.cc:329]: Saving the flow information to the edgelabel. If we have one of the flow sources: kFreeFlowMask, kConstrainedFlowMask, kPredictedFlowMask, or kCurrentFlowMask then we have flow data and set the bit to true.

File: src/thor/bidirectional_astar.cc
Comment (gknisely) [inline src/thor/bidirectional_astar.cc:522]: Added flow_sources to TransitionCostReverse so that we can save it off in the edge label.

File: src/thor/bidirectional_astar.cc
Comment (gknisely) [inline src/thor/bidirectional_astar.cc:291]: Added flow_sources to EdgeCost so that we can save it off in the edge label.

File: valhalla/sif/edgelabel.h
Comment (gknisely) [inline valhalla/sif/edgelabel.h:428]: Our new bit for flow.

File: src/sif/autocost.cc
Comment (mandeepsandhu) [inline src/sif/autocost.cc:369]: dumb q: whats the reasoning for the math here (why the cube)?

File: valhalla/sif/edgelabel.h
Comment (kevinkreiser) [reply valhalla/sif/edgelabel.h:428]: i would suggest we name that `has_measured_speed_` since we really want this for any edge that has measured traffic speed (flow/constrained/historical/realtime)

File: src/sif/autocost.cc
Comment (kevinkreiser) [reply src/sif/autocost.cc:369]: obviously because @gknisely only has 3 F's to give! badum-cha!

Comment (gknisely) [reply src/sif/autocost.cc:369]: copied from motorcycle costing.  we were seeing better results.  @dnesbitt61 originally wrote this logic.  Do you know why you did this?

Comment (dnesbitt61) [reply src/sif/autocost.cc:369]: some secret sauce likely arrived at doing some route testing with motorcycles routes to see what gave results that seemed good. Luckily I commented my approach:
  // Factor for trail use - use a non-linear factor with values at 0.5 being neutral (factor
  // of 0). Values between 0.5 and 1 slowly decrease to a maximum of -0.125 (to slightly prefer
  // trails) while values between 0.5 to 0 slowly increase to a maximum of the surfact_factor_
  // to avoid/penalize trails.
Now this was applied to motorcycles where there were 2 different route case: first being for road motorcycling and second more adventure motorcyling. So I cannot vouch for its use in auto costing

File: src/sif/autocost.cc
Comment (dgearhart) [inline src/sif/autocost.cc:606]: This is where reduced stop impact cost when using measured speed

File: src/sif/autocost.cc
Comment (dgearhart) [inline src/sif/autocost.cc:545]: This is where reduced stop impact cost when using measured speed

Comment (dgearhart) [reply src/sif/autocost.cc:545]: More work to be done in the future to balance cost with measured speed

File: valhalla/sif/edgelabel.h
Comment (gknisely) [reply valhalla/sif/edgelabel.h:429]: @kevinkreiser updated

File: run_route_scripts/batch.sh
Comment (dgearhart) [inline run_route_scripts/batch.sh:61]: Header updates fro elapsed cost statistics

File: run_route_scripts/results/combine_route_stats.py
Comment (dgearhart) [inline run_route_scripts/results/combine_route_stats.py:8]: enable elapsed cost stats for delta and % delta of baseline vs latest

File: src/valhalla_run_route.cc
Comment (dgearhart) [inline src/valhalla_run_route.cc:74]: add elapsed cost into stats

**Review by kdiluca — APPROVED** [review]
Knowing that the RAD results were good for this then I approve however we need to add more tests!

File: src/sif/autocost.cc
Comment (mandeepsandhu) [inline src/sif/autocost.cc:608]: Why don't we use the predecessor label's `has_measure_speed()` instead of passing-in the value explicitly here (similar to whats done in `TransitionCost()` ?

Comment (dnesbitt61) [reply src/sif/autocost.cc:608]: the predecessor EdgeLabel is in reverse direction of actual travel. I imagine the 2 directed edges in opposing directions could have different speeds / measured speed flags.

File: src/sif/autocost.cc
Comment (mandeepsandhu) [reply src/sif/autocost.cc:369]: Yeah, I guess RAD tests will show if this is generating good results for auto or not (which I believe is good, right @gknisely ?)

File: src/sif/autocost.cc
Comment (mandeepsandhu) [reply src/sif/autocost.cc:454]: Any reason why this isn't a pointer like the equivalent param in `GraphTile::GetSpeed()`? Pointer is good if not everyone is expected to use this param (eg in `DynamicCost::EdgeCost` we're creating an unused var because of the need for pass by ref).

File: src/sif/autocost.cc
Comment (mandeepsandhu) [reply src/sif/autocost.cc:608]: ah ok, that makes sense. thx @dnesbitt61

File: src/sif/motorcyclecost.cc
Comment (mandeepsandhu) [inline src/sif/motorcyclecost.cc:256]: Is `has_measured_speed` unused here?

File: src/sif/motorscootercost.cc
Comment (mandeepsandhu) [inline src/sif/motorscootercost.cc:296]: Is `has_measured_speed` unsed here? If so, maybe make it an unnamed param to avoid compiler warning.

Comment (mandeepsandhu) [reply src/sif/motorscootercost.cc:296]: Same comment for all other places where we have this param unused.

File: src/sif/autocost.cc
Comment (gknisely) [reply src/sif/autocost.cc:608]: @mandeepsandhu due to this https://github.com/valhalla/valhalla/issues/2896

File: src/sif/motorcyclecost.cc
Comment (gknisely) [reply src/sif/motorcyclecost.cc:256]: @mandeepsandhu we are missing RAD tests for this and it is noted as a TODO

Comment (mandeepsandhu) [reply src/sif/motorcyclecost.cc:256]: ignore this comment, I didn't realize this was the function declaration.

File: src/sif/motorscootercost.cc
Comment (mandeepsandhu) [reply src/sif/motorscootercost.cc:296]: ignore these too. sorry about the noise!

Comment (gknisely) [reply src/sif/motorscootercost.cc]: @mandeepsandhu updated

File: valhalla/sif/dynamiccost.h
Comment (mandeepsandhu) [inline valhalla/sif/dynamiccost.h:911]: Why are we adding `service_penalty_` twice?

File: valhalla/sif/edgelabel.h
Comment (mandeepsandhu) [reply valhalla/sif/edgelabel.h:429]: no more spare bits left :slightly_frowning_face:

**Comment by mandeepsandhu** [comment]
> Knowing that the RAD results were good for this then I approve however we need to add more tests!

:100:  we need more tests!

File: valhalla/sif/dynamiccost.h
Comment (gknisely) [reply valhalla/sif/dynamiccost.h]: @mandeepsandhu merge issue.  Thank you for seeing this one

Comment (gknisely) [reply valhalla/sif/dynamiccost.h]: updated.

Comment (mandeepsandhu) [reply valhalla/sif/dynamiccost.h]: Since this affected the base transition cost, should we RAD test a few routes to see the result is still as expected? Although if this was only due to a recent merge and the RAD results in PR were performed before this change, then we can skip it. I'll leave it to you

**Review by mandeepsandhu — APPROVED** [review]
Changes look good. Approving assuming RAD with latest changes look good as well.

File: src/sif/autocost.cc
Comment (PureW) [reply src/sif/autocost.cc:369]: @gknisely Can you copy that code comment and modify if necessary to this code? I`m with @mandeepsandhu here :)

File: src/thor/bidirectional_astar.cc
Comment (genadz) [reply src/thor/bidirectional_astar.cc:329]:
```c++
kFreeFlowMask | kConstrainedFlowMask | kPredictedFlowMask | kCurrentFlowMask
```
- this is used several times, probably would be better to change it with a new constants ? `kFlowMask`

Comment (PureW) [reply src/thor/bidirectional_astar.cc:329]: Yes, plus one on this

Comment (kevinkreiser) [reply src/thor/bidirectional_astar.cc:329]: we have that already: `kDefaultFlowMask`

File: valhalla/sif/edgelabel.h
Comment (kevinkreiser) [reply valhalla/sif/edgelabel.h:429]: there are 17 more spare below. there just arent anymore spare in _this_ 64bit value

File: src/thor/bidirectional_astar.cc
Comment (gknisely) [reply src/thor/bidirectional_astar.cc:327]: @genadz @PureW @kevinkreiser thank you i missed that one

**Review by mandeepsandhu — APPROVED** [review]

**Pull request merged**
